semantics/cpp14/language/translation/decl/enum.k

```diff
@@ -0,0 +1,218 @@
+
+     syntax Enum ::= createEnumName(CId, Scope) [function]
+     rule createEnumName(X::CId, namespaceScope(N::Namespace)) => N :: Enum(X)
+     rule createEnumName(X::CId, classScope(C::Class)) => C :: Enum(X)
+     rule createEnumName(X::CId, blockScope(_, _, _) #as B::BlockScope) => localQual(B) :: Enum(X)
+
+     syntax CPPEnumType ::= declareEnumName(Enum, ut: CPPIntegerType, scoped: Bool, fixed: Bool) [function]
+     rule declareEnumName(E::Enum, UT::CPPIntegerType, false, Fixed:Bool) => t(noQuals, .Set, unscopedEnum(enumId(E), UT, false))
+     rule declareEnumName(E::Enum, UT::CPPIntegerType, true, true)  => t(noQuals, .Set, scopedEnum(enumId(E), UT))
+
+     syntax Scope ::= getEnumScope(EnumId) [function]
+     rule getEnumScope(enumId(N:Namespace :: _)) => namespaceScope(N)
+
+     syntax KItem ::= addEnumToEnv(CId, CPPEnumType)
+     rule addEnumToEnv(X::CId, E::CPPEnumType) => #addEnumToEnv(X, getEnumScope(getEnumId(E)), E)
+
+     syntax KItem ::= #addEnumToEnv(CId, Scope, CPPEnumType)
+     rule <k> #addEnumToEnv(X::CId, namespaceScope(N::Namespace), E::CPPEnumType) => . ... </k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          <ns-id> N </ns-id>
+          <ntypes> NT::Map => NT[X, Enum() <- E] </ntypes>
+
+     syntax KItem ::= enumContext(CId, CPPEnumType, List, fixed: Bool, ut: CPPIntegerType) [strict(2)]
+
+     rule enumContext(X::CId, E::CPPEnumType, Enumerators:List, Fixed::Bool, UT::CPPIntegerType)
+          => newEnum(E, Fixed)
+          ~> addEnumToEnv(X, E)
+          ~> scope(enumScope(getEnumId(E)), declareEnumerators(E, 0, 0, 0, int, Enumerators, Fixed, true))
+
+     // second param -underlying type
+     syntax KItem ::= newEnum(CPPEnumType, fixed: Bool)
+
+     rule <k> newEnum(E:CPPEnumType, true) => . ... </k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          (.Bag => <cppenum>...
+               <enum-id> getEnumId(E) </enum-id>
+               <enum-type> E </enum-type>
+               <scoped> getScoped(E) </scoped>
+               <enum-complete> true </enum-complete>
+               <enum-min> min(underlyingType(E)) </enum-min>
+               <enum-max> max(underlyingType(E)) </enum-max>
+           ...</cppenum>)
+
+     rule <k> newEnum(E:CPPUnscopedEnumType, false) => . ... </k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          (.Bag => <cppenum>...
+               <enum-id> getEnumId(E) </enum-id>
+               <enum-type> E </enum-type>
+               <scoped> false </scoped>
+               <enum-complete> false </enum-complete>
+           ...</cppenum>)
+
+     syntax KItem ::= declareEnumerators(CPPEnumType, min: Int, max: Int, next: Int, nextType: CPPSimpleType, List, fixed: Bool, first: Bool)
+
+     rule declareEnumerators(_, _, _, Next::Int, NextType::CPPSimpleType, ListItem(Enumerator(_, NoExpression() => prv(Next, noTrace, type(NextType)))) _::List, _, _)
```

Comment:
There is no UNSPEC in C++ semantic, so I will take the definition from C semantics. I guess that I should do the rewrite even if it is unspecified, so that rv-match could recover from that. So the rule will look like
`rule <k> (.K => UNSPEC(...)) ~> a(b=>c, true)` ... </k>